### DISCUSSION TITLE
How to get an image object and update this image in rust

### DISCUSSION THREAD

**Discussion by CarbonPool** [slint-ui/slint#2527] [Q&A]
Appreciate any help, I didn't find relevant examples

**Comment by ghost** [comment]
You can found descriptions and examples on the rust documentation: https://slint-ui.com/releases/1.0.0/docs/rust/slint/struct.image. If you need more input feel free to ask.

**Comment by ogoffart** [comment]
If you have a property of type `image`,  this is goinf to be a `slint::Image`  which you can replace.

But what exactly do you mean by "updating" an image?

**Reply by Martoni** [reply]
I have the same problem in [my test project](https://github.com/Martoni/QRNote/blob/main/qr-note/ui/appwindow.slint).

A QRcode image is displayed on main windows : 
```rust
export component AppWindow inherits Window {
    in-out property<int> counter: 42;
    in-out property<image> qrnote: @image-url("icons/QRNote.png");
    callback compute-qr-code();
//...
        Button {
            text: "Gen QRNote";
            clicked => {
                root.compute-qr-code();
            }
        }
        Image {
            source : qrnote;
            width: 100px;
            height: 100px;
        }
    }
}
```

But I want to refresh it with a new generated QRCode each time I click on a button : 

```rust
//...
    let ui_handle = ui.as_weak();
    ui.on_compute_qr_code(move || {
        let ui = ui_handle.unwrap();
        let code = QrCode::new(ui.get_thetext()).unwrap();
        let image = code.render::<Luma<u8>>().build();
        image.save("ui/icons/qrcode.png").unwrap();
        use slint::Image;
        let qrimage = Image::load_from_path(Path::new("ui/icons/qrcode.png")).unwrap();
        ui.set_qrnote(qrimage); // <- Doesn't refresh image !
    });
```
When the callback `on_compute_qr_code is called, image is not refreshed.
It seems to be the same problem

**Reply by ogoffart** [reply]
I think this is a bug.
We have a cache and and we don't reload the image if it was made from the same path.

https://github.com/slint-ui/slint/blob/25b380bcc82ab9bb178af5dbe5187c6909cb42f0/internal/core/graphics/image.rs#L578-L581

@tronical should we always reload the image? or should we perhaps check the modtime of the file?

**Reply by Martoni** [reply]
Anyway, the way I do it is little bit overkill.
Saving the generated image on disk and re-load it is certainly not the best method.
But I'm learning Slint (and rust !) and I haven't found the right method yet.

**Reply by tronical** [reply]
Indeed, that's the bug. Suppose we have this in a .slint file:

```
Image {
    source: @image-url("foo.png");
}
Image {
    source: @image-url("foo.png");
}
```

Ideally we should load `foo.png` only once from disk. A modification time check seems indeed reasonable. The only other option I can think of is that the above .slint code would generate rust/c++ code that uses a different function (that doesn't check the modification time) and we make `Image::load_from_path` /always/ load from disk. But I don't quite like that either, so an extra stat() seems like the best option to me.

Sounds ok to you @ogoffart ?

**Reply by ogoffart** [reply]
True, you shouldn't need to save it on the disc, you can use [slint::image::from_rgb8](https://docs.rs/slint/latest/slint/struct.Image.html#method.from_rgb8) or one of the other similar function.  And you can probably find a way to convert your image to a slint::SharedPixelBuffer.   the slint::Image documentation has some examples.

**Reply by Martoni** [reply]
Thanks @ogoffart , I find the way to do it properly ;)
https://github.com/Martoni/QRNote/blob/main/qr-note/src/main.rs
